File: lib/simulator/src/math.ts
/**
 * mathe:buddy - a gamified learning-app for higher math
 * (c) 2022-2023 by TH Koeln
 * Author: Andreas Schwenk contact@compiler-construction.com
 * Funded by: FREIRAUM 2022, Stiftung Innovation in der Hochschullehre
 * License: GPL-3.0-or-later
 */

// TODO: only use mathebuddy-math-runtime for math!!

/*
export class MathIntSet {
  private values: number[] = [];

  //G intSet = "{" { INT } "}";
  parse(s: string): boolean {
    s = s.trim().replace('/ /g', '');
    if (s.length < 2) return false;
    if (s[0] != '{' || s[s.length - 1] != '}') return false;
    s = s.substring(1, s.length - 1);
    const v = s.split(',');
    for (const vi of v) this.values.push(parseInt(vi));
    console.log(this.values);
    return true;
  }

  compare(s: MathIntSet): boolean {
    if (this.values.length != s.values.length) return false;
    for (const u of this.values) {
      let found = false;
      for (const v of s.values) {
        if (u == v) {
          found = true;
          break;
        }
      }
      if (!found) return false;
    }
    return true;
  }
}

export function compareIntSets(a: string, b: string): boolean {
  const setA = new MathIntSet();
  if (setA.parse(a) == false) return false;
  const setB = new MathIntSet();
  if (setB.parse(b) == false) return false;
  return setA.compare(setB);
}
*/

File: lib/compiler/src/math.dart
/// MatheBuddy - a gamified app for higher math
/// https://mathebuddy.github.io/
/// (c) 2022-2024 by TH Koeln
/// Author: Andreas Schwenk contact@compiler-construction.com
/// Funded by: FREIRAUM 2022, Stiftung Innovation in der Hochschullehre
/// License: GPL-3.0-or-later

library mathe_buddy_compiler;

import 'package:slex/slex.dart';

import '../../mbcl/src/level.dart';
import '../../mbcl/src/level_item.dart';

/// <GRAMMAR>
///   inlineMathCore =
///     { "term" "(" ID ")" | "term" "(" ID ")" | ID } .."$";
/// </GRAMMAR>
MbclLevelItem parseInlineMath(
    MbclLevel level, Lexer lexer, MbclLevelItem? exercise) {
  var lastTk = "";
  if (lexer.isTerminal('\$')) lexer.next();
  var inlineMath = MbclLevelItem(level, MbclLevelItemType.inlineMath, -1);
  while (lexer.isNotTerminal('\$') && lexer.isNotEnd()) {
    var tk = lexer.getToken().token;
    // ID | "opt" "(" ID ")" | "term" "(" ID ")";
    var showTerm = false;
    var showOptimizedTerm = false;
    var id = lexer.getToken().token;
    var isId = lexer.getToken().type == LexerTokenType.id;
    if (lexer.isTerminal("term")) {
      showTerm = true;
    } else if (lexer.isTerminal("opt")) {
      showOptimizedTerm = true;
    }
    lexer.next();
    if (showTerm || showOptimizedTerm) {
      if (lexer.isTerminal("(")) lexer.next();
      id = lexer.getToken().token;
      isId = lexer.getToken().type == LexerTokenType.id;
      lexer.next();
      if (lexer.isTerminal(")")) lexer.next();
    }
    // variable reference ?
    if (isId && exercise != null) {
      var data = exercise.exerciseData!;
      if (data.variables.contains(id)) {
        var type = MbclLevelItemType.variableReferenceOperand;
        if (showTerm || data.functionVariables.contains(id)) {
          type = MbclLevelItemType.variableReferenceTerm;
        }
        if (showOptimizedTerm) {
          type = MbclLevelItemType.variableReferenceOptimizedTerm;
        }
        var v = MbclLevelItem(level, type, -1);
        v.id = id;
        inlineMath.items.add(v);
        continue;
      }
    }
    // default (no variable reference)
    var text = MbclLevelItem(level, MbclLevelItemType.text, -1);
    if (lastTk == "\\") {
      text.text = tk;
    } else {
      text.text = " " + tk;
    }
    inlineMath.items.add(text);
    lastTk = tk;
  }
  if (lexer.isTerminal('\$')) lexer.next();
  return inlineMath;
}
